src/tasks.jl

```diff
@@ -0,0 +1,233 @@
+# This way, we do no longer need to have the model and the sampler saved in the trace struct
+# However, we still need to insantiate the VarInfo
+
+# TaskInfo stores additional informatino about the task
+mutable struct Trace{Tvi, TInfo}  <: AbstractTrace where {Tvi, TInfo <: AbstractTaskInfo}
+    vi::Tvi # Unfortunatley, we can not set force this to be a subtype of VarInfo...
+    task::Task
+    taskinfo::TInfo
+end
+
+function Base.copy(trace::Trace{Tvi,TInfo}, copy_vi::Function) where {Tvi, TInfo <: AbstractTaskInfo}
+    res = Trace{typeof(trace.vi),typeof(trace.taskinfo)}(copy_vi(trace.vi), copy(trace.task), copy(trace.taskinfo))
+    return res
+end
+
+# The procedure passes a function which is specified by the model.
+
+function Trace(f::Function, vi, taskinfo, copy_vi::Function)
+    task = CTask( () -> begin res=f(); produce(Val{:done}); res; end )
+
+    res = Trace{typeof(vi),typeof(task)}(copy_vi(vi), task, copy(taskinfo));
+    # CTask(()->f());
+    if res.task.storage === nothing
+        res.task.storage = IdDict()
+    end
+    res.task.storage[:turing_trace] = res # create a backward reference in task_local_storage
+    return res
+end
+
+## We need to design the task in the Turing wrapper.
+function Trace( vi, task::Task, taskinfo, copy_vi::Function)
+
+
+    res = Trace{typeof(vi),typeof(taskinfo)}(copy_vi(vi), Libtask.copy(task), copy(taskinfo));
+    # CTask(()->f());
+    if res.task.storage === nothing
+        res.task.storage = IdDict()
+    end
+    res.task.storage[:turing_trace] = res # create a backward reference in task_local_storage
+    return res
+end
+
+
+
+# step to the next observe statement, return log likelihood
+Libtask.consume(t::Trace) = (t.vi.num_produce += 1; consume(t.task))
+
+# Task copying version of fork for Trace.
+function fork(trace :: Trace, copy_vi::Function, is_ref :: Bool = false, set_retained_vns_del_by_spl!::Union{Function,Nothing} = nothing)
+    newtrace = copy(trace,copy_vi)
+    if is_ref
+        @assert set_retained_vns_del_by_spl! != nothing "[AdvancedPF] set_retained_vns_del_by_spl! is not set."
+        set_retained_vns_del_by_spl!(newtrace.vi)
+    end
+    newtrace.task.storage[:turing_trace] = newtrace
+    return newtrace
+end
+
+# PG requires keeping all randomness for the reference particle
+# Create new task and copy randomness
+function forkr(trace :: Trace, copy_vi::Function)
+
+    newtrace = Trace(trace.vi,trace.task ,trace.taskinfo,copy_vi)
+    newtrace.vi.num_produce = 0
+    return newtrace
+end
+
+current_trace() = current_task().storage[:turing_trace]
+
+const Particle = Trace
+
+"""
+Data structure for particle filters
+- effectiveSampleSize(pc :: ParticleContainer)
+- normalise!(pc::ParticleContainer)
+- consume(pc::ParticleContainer): return incremental likelihood
+"""
+mutable struct ParticleContainer{Tvi,TInfo} <: AbstractParticleContainer where {Tvi, TInfo <: AbstractTaskInfo}
+    vals::Vector{Trace{Tvi,TInfo}}
```

Comment:
Is it necessary to be so explicit with the types? Couldn't you just define
```julia
mutable struct ParticleContainer{T<:Trace} <: AbstractParticleContainer
    vals::Vector{T}
```
And I don't see immediately why the additional type parameters would be useful?

Reply:
Good point, however, this implies that we need to define the Trace in Turing. One of the goals is that we can make changes in this package without the need for making corrections in Turing.

Reply:
Can you explain why that would imply that you have to define `Trace` in Turing? I don't see immediately why that would be the case.

Reply:
I assume your point is that Turing needs  a slightly different `Trace`? Then I would suggest
```julia
mutable struct ParticleContainer{T} <: AbstractParticleContainer
    vals::Vector{T}
```
That could be used with arbitrary particle types.

Reply:
Thanks! I will change it